Sentieon docs (#1131)
nf-core/sarek: docs/output.md

These files are intermediate and by default not placed in the output-folder kept in the final files delivered to users. Set `--save_mapped` to enable publishing, furthermore add the flag `save_output_as_bam` for publishing in BAM format.

#### Sentieon BWA mem

Sentieon [bwa mem](https://support.sentieon.com/manual/usages/general/#bwa-mem-syntax) is a subroutine for mapping low-divergent sequences against a large reference genome. It is part of the proprietary software package [DNAseq](https://www.sentieon.com/detailed-description-of-pipelines/#dnaseq) from [Sentieon](https://www.sentieon.com/).

The aligned reads are coordinate-sorted with Sentieon.

<details markdown="1">
<summary>Output files for all mappers and samples</summary>

> The alignment files (BAM or CRAM) produced by the chosen aligner are, by default, not published, that is, they are not placed in the output-folder (`outdir`), but by setting `--save_mapped` the alignment files are published in CRAM format or, by additional setting `--save_output_as_bam`, in BAM format.

Reviewer: Does one really have to additionally set the `--save_output_as_bam` to output CRAM and BAM? I would have thought it's either CRAM or BAM.
```suggestion
The alignment files (BAM or CRAM) produced by the chosen aligner are not published by default. CRAM output files will not be saved in the output-folder (`outdir`), unless the flag --save_mapped` is used. BAM output can be selected by setting the flag `--save_output_as_bam`.
```

Author: As I undertand it, the alignment files directly from the aligner are not saved by default (BAM or CRAM may be saved from some later step like MarkDuplicatess). To save CRAM-files from directly from the aligner one uses `--save_mapped`, and, if one wants BAMs instead of CRAMS one adds `--save_output_as_bam`.

I'd accept your suggestion, but I think there is a quotation mark missing around `--save_mapped`. Could you fix that, then I can subsequently accept.

Reviewer: Sorry, should be there now

**Output directory: `{outdir}/preprocessing/mapped/<sample>/`**

> - if `--save_mapped`: `<sample>.sorted.cram` and `<sample>.sorted.cram.crai`

Reviewer 2: Isn't it that depending on tools used, I'm thinking especially markduplicates and or spark, the files are sorted and not just aligned.
Maybe let's drop the `.sorted` or `.mapped` here in the docs, and people will figure it out, no?

> - if `--save_mapped`: `<sample>.sorted.cram` and `<sample>.sorted.cram.crai`

Author: As far as I can tell, no process outputs anything with the substring `.mapped` or `mapped.`. MarkDuplicate (with or without spark) outputs files with the substring `.md.`, Sentieon Dedup outputs files with the substring `.dedup.`.

On the `master`-branch, I checked that the BAM and CRAM from the following command are sorted:
```
nextflow run main.nf -profile test_cache,docker --save_mapped --save_output_as_bam --outdir results
```
that is, `results/preprocessing/mapped/test/test.sorted.bam` is coordinate-sorted. (That command runs `BWAMEM1_MEM` and that coordinate sorting.)

Similarly, I ran
```
nextflow run main.nf -profile test_cache,docker --save_mapped --outdir results
```
and found that `results/preprocessing/mapped/test/test.sorted.cram` is coordinate-sorted. The CRAM-file `test.sorted.cram` is actually published by `BAM_TO_CRAM_MAPPING` (`SAMTOOLS_CONVERT`), but that just converts from BAM to CRAM and that doesn't ruin the sorting of the reads.

The other aligners (including SENTIEON_BWAMEM) also does sorting.

> - if `--save_mapped`: `<sample>.sorted.cram` and `<sample>.sorted.cram.crai`

Reviewer 2: yeah, then for me `sorted` should trump `mapped`.
So ok for me to keep `sorted`

</details>

### Sentieon LocusCollector and Dedup

The subroutines LocusCollector and Dedup are part of Sentieon DNAseq packages with speedup versions of the standard GATK tools, and together those two subroutines correspond to GATK's MarkDuplicates.

> The subroutine [LocusCollector](https://support.sentieon.com/manual/usages/general/#driver-algorithm-syntax) collects read information that will be used for removing or marking of duplicate reads; its output is the score file indicating which reads are likely duplicates.

Reviewer: ```suggestion
The subroutine [LocusCollector](https://support.sentieon.com/manual/usages/general/#driver-algorithm-syntax) collects read information that will be used for removing or tagging duplicate reads; its output is the score file indicating which reads are likely duplicates.
```

> The subroutine [Dedup](https://support.sentieon.com/manual/usages/general/#dedup-algorithm) marks or removes duplicate reads based no the score file supplied by LocusCollector, and produces a BAM or CRAM file.

Reviewer: ```suggestion
The subroutine [Dedup](https://support.sentieon.com/manual/usages/general/#dedup-algorithm) marks or removes duplicate reads based on the score file supplied by LocusCollector, and produces a BAM or CRAM file.
```

[GATK Joint germline Variant Calling](https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-) uses Haplotypecaller per sample in `gvcf` mode. Next, the gVCFs are consolidated from multiple samples into a [GenomicsDB](https://gatk.broadinstitute.org/hc/en-us/articles/5358869876891-GenomicsDBImport) datastore. After joint [genotyping](https://gatk.broadinstitute.org/hc/en-us/articles/5358906861083-GenotypeGVCFs), [VQSR](https://gatk.broadinstitute.org/hc/en-us/articles/5358906115227-VariantRecalibrator) is applied for filtering to produce the final multisample callset with the desired balance of precision and sensitivity.

<details markdown="1">

> <summary>Output files from joint germline variant callling</summary>

Reviewer: ```suggestion
<summary>Output files from joint germline variant calling</summary>
```

Author: I need to get my keyboard fixed 🤦‍♂️

**Output directory: `{outdir}/variantcalling/haplotypecaller/<sample>/`**

> - `<sample>.haplotypecaller.g.vcf.gz` and `<sample>.haplotypecaller.g.vcf.gz.tbi`

Reviewer: Should this `.g.` really be there?

Author: yes

- `{sample,tumorsample_vs_normalsample}.mutect2.filtered.vcf.gz.filteringStats.tsv`
  - a stats file generated during the filtering of Mutect2 called variants

> </details

Reviewer 2: Not sure about using this closing tag here

</details>

#### Sentieon Haplotyper

[Sentieon Haplotyper](https://support.sentieon.com/manual/usages/general/#haplotyper-algorithm) is Sention's speedup version of GATK's Haplotypecaller (see above).

<details markdown="1">
<summary>Unfiltered VCF-files for normal samples</summary>

**Output directory: `{outdir}/variantcalling/sentieon_haplotyper/<sample>/`**

- `<sample>.haplotyper.unfiltered.vcf.gz` and `<sample>.haplotyper.unfiltered.vcf.gz.tbi`
  - VCF with tabix index

</details>

The output from Sentieon's Haplotyper can be controlled through the option `--sentieon_haplotyper_emit_mode` for Sarek, see [Basic usage of Sentieon functions in Sarek](https://github.com/nf-core/sarek/blob/sentieon_docs/docs/usage.md#basic-usage-of-sentieon-functions-in-sarek).

> Unless `haplotyper_filter` is listed under `--skip_tools` in the nextflow command, GATK's CNNScoreVariants and FilterVariantTranches (see above) is applied to the unfiltered VCF-files in order to obtained filtered vcf-files.

Reviewer: ```suggestion
Unless `haplotyper_filter` is listed under `--skip_tools` in the nextflow command, GATK's CNNScoreVariants and FilterVariantTranches (see above) is applied to the unfiltered VCF-files in order to obtain filtered vcf-files.
```

Reviewer: either `vcf` or `VCF`

<details markdown="1">
<summary>Filtered VCF-files for normal samples</summary>

**Output directory: `{outdir}/variantcalling/sentieon_haplotyper/<sample>/`**

- `<sample>.haplotyper.filtered.vcf.gz` and `<sample>.haplotyper.filtered.vcf.gz.tbi`
  - VCF with tabix index

</details>

##### Sentieon Joint Germline Variant Calling

> In Sentieon's package DNAseq, joint germline variant calling is done by first running Sentieon's Haplotyper in emit-mode `gvcf` for each sample, and then running Sentieon's [GVCFtyper](https://support.sentieon.com/manual/usages/general/#gvcftyper-algorithm) on the set of gVCF-files. See [Basic usage of Sentieon functions in Sarek](https://github.com/nf-core/sarek/blob/sentieon_docs/docs/usage.md#basic-usage-of-sentieon-functions-in-sarek) for information on how joint germline variant callling can be done in Sarek using Sentieon's DNAseq.

Reviewer: ```suggestion
In Sentieon's package DNAseq, joint germline variant calling is done by first running Sentieon's Haplotyper in emit-mode `gvcf` for each sample and then running Sentieon's [GVCFtyper](https://support.sentieon.com/manual/usages/general/#gvcftyper-algorithm) on the set of gVCF-files. See [Basic usage of Sentieon functions in Sarek](https://github.com/nf-core/sarek/blob/sentieon_docs/docs/usage.md#basic-usage-of-sentieon-functions-in-sarek) for information on how joint germline variant calling can be done in Sarek using Sentieon's DNAseq.
```

> Sarek's implementation of joint germline variant calling using DNAseq does not include the usage of [GenomicsDB](https://gatk.broadinstitute.org/hc/en-us/articles/5358869876891-GenomicsDBImport) datastore. After joint genotyping, Sentieon's version of VQSR ([VarCal](https://support.sentieon.com/manual/usages/general/#varcal-algorithm) and [ApplyVarCal](https://support.sentieon.com/manual/usages/general/#applyvarcal-algorithm)) is applied for filtering to produce the final multisample callset with the desired balance of precision and sensitivity.

Reviewer: ```suggestion
Sarek's implementation of joint germline variant calling using DNAseq does not include the usage of [GenomicsDB](https://gatk.broadinstitute.org/hc/en-us/articles/5358869876891-GenomicsDBImport) datastore. After joint genotyping, Sentieon's version of VQSR ([VarCal](https://support.sentieon.com/manual/usages/general/#varcal-algorithm) and [ApplyVarCal](https://support.sentieon.com/manual/usages/general/#applyvarcal-algorithm) is applied for filtering to produce the final multisample callset with the desired balance of precision and sensitivity.
```

<details markdown="1">

> <summary>Output files from joint germline variant callling</summary>

Reviewer: :D 
```suggestion
<summary>Output files from joint germline variant calling</summary>
```

**Output directory: `{outdir}/variantcalling/sentieon_haplotyper/<sample>/`**

> - `<sample>.haplotypecaller.g.vcf.gz` and `<sample>.haplotypecaller.g.vcf.gz.tbi`

Reviewer: same question as above

Author: Yes, a gVCF-file is created for each sample in the joint-germline workflow. In Sarek, they get the extension `g.vcf`.

  - file used by [MultiQC](https://multiqc.info/)
  </details>

#### Sentieon Dedup reports

> Sentieon's DNAseq subroutine Dedup produces a metrics report much like the one produce by GATK's MarkDuplicates. The Dedup metrics are imported into MultiQC as custom content and displayed in a table.

Reviewer: ```suggestion
Sentieon's DNAseq subroutine Dedup produces a metrics report much like the one produced by GATK's MarkDuplicates. The Dedup metrics are imported into MultiQC as custom content and displayed in a table.
```
